OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/ifds/problems/VariableTypeProblem.scala

```diff
@@ -0,0 +1,375 @@
+/* BSD 2-Clause License - see OPAL/LICENSE for details. */
+package org.opalj.tac.fpcf.analyses.ifds.problems
+
+import org.opalj.br.ArrayType
+import org.opalj.br.ClassFile
+import org.opalj.br.FieldType
+import org.opalj.br.Method
+import org.opalj.br.ReferenceType
+import org.opalj.br.analyses.DeclaredMethods
+import org.opalj.br.analyses.DeclaredMethodsKey
+import org.opalj.br.analyses.SomeProject
+import org.opalj.br.fpcf.PropertyStoreKey
+import org.opalj.collection.immutable.EmptyIntTrieSet
+import org.opalj.fpcf.FinalEP
+import org.opalj.fpcf.PropertyStore
+import org.opalj.ifds.AbstractIFDSFact
+import org.opalj.ifds.AbstractIFDSNullFact
+import org.opalj.ifds.Callable
+import org.opalj.ifds.Dependees.Getter
+import org.opalj.ifds.IFDSFact
+import org.opalj.tac.ArrayLoad
+import org.opalj.tac.ArrayStore
+import org.opalj.tac.Assignment
+import org.opalj.tac.DUVar
+import org.opalj.tac.Expr
+import org.opalj.tac.GetField
+import org.opalj.tac.GetStatic
+import org.opalj.tac.New
+import org.opalj.tac.ReturnValue
+import org.opalj.tac.Var
+import org.opalj.tac.fpcf.analyses.ifds.JavaForwardIFDSProblem
+import org.opalj.tac.fpcf.analyses.ifds.JavaIFDSProblem
+import org.opalj.tac.fpcf.analyses.ifds.JavaStatement
+import org.opalj.tac.fpcf.properties.cg.Callers
+import org.opalj.value.ValueInformation
+
+import scala.annotation.tailrec
+
+trait VTAFact extends AbstractIFDSFact
+case object VTANullFact extends VTAFact with AbstractIFDSNullFact
+
+/**
+ * A possible run time type of a variable.
+ *
+ * @param definedBy The variable's definition site.
+ * @param t The variable's type.
+ * @param upperBound True, if the variable's type could also be every subtype of `t`.
+ */
+case class VariableType(definedBy: Int, t: ReferenceType, upperBound: Boolean) extends VTAFact {
+
+    /**
+     * If this VariableType is an upper bound, it subsumes every subtype.
+     */
+    override def subsumes(other: AbstractIFDSFact, project: SomeProject): Boolean = {
+        if (upperBound) other match {
+            case VariableType(definedByOther, tOther, _) if definedBy == definedByOther && t.isObjectType && tOther.isObjectType =>
```

Comment:
Is there a reason these are checked to be object types? The class hierarchy can do subtype checks also for general reference types. There might be a good reason to do this here, though, so this should be checked.